src/logic.rs

```diff
@@ -8,141 +6,112 @@ use cardano_serialization_lib::metadata::{
 
-    for (rego, voting_power) in rego_voting_power {
-        let entry = Output {
-            delegations: rego.metadata.delegations.clone(),
-            rewards_address: rego.metadata.rewards_addr.clone(),
-            stake_public_key: rego.metadata.stake_vkey.clone().convert(),
-            voting_power: voting_power.into(),
-            voting_purpose: rego.metadata.purpose,
-        };
-        output.push(entry);
-    }
+            let voting_power = values.get(&stake_addr as &str)?.clone();
 
-    Ok(output)
-}
+            Some(Output {
+                delegations: reg.metadata.delegations.clone(),
+                rewards_address: reg.metadata.rewards_addr.clone(),
+                stake_public_key: reg.metadata.stake_vkey.convert(),
+                voting_purpose: reg.metadata.purpose,
+                voting_power,
+            })
+        })
+        .collect();
 
-fn filter_latest_registrations(regos: impl IntoIterator<Item = Rego>) -> Vec<Rego> {
-    // Group the registrations by stake key (each stake key may have one valid registration)
-    let mut m = HashMap::new();
-    for rego in regos {
-        let stake_key = rego.metadata.stake_vkey.clone();
-        m.entry(stake_key).or_insert_with(Vec::new).push(rego);
-    }
-    // Find the regos with the highest slot number, and of those, choose the
-    // lowest txid.
-    let mut latest_regos = Vec::new();
-    for (_, stake_regos) in m {
-        let latest = stake_regos
-            .iter()
-            .fold(stake_regos[0].clone(), |acc, rego| {
-                use std::cmp::Ordering::*;
-                match &rego.metadata.slot.cmp(&acc.metadata.slot) {
-                    // If the slot number is less, it's not a newer registration.
-                    Less => acc,
-                    // If the slot number is greater, it's a newer registration.
-                    Greater => rego.clone(),
-                    // If the slot number is equal, choose the one with the lower tx id.
-                    Equal => {
-                        if rego.tx_id < acc.tx_id {
-                            rego.clone()
-                        } else {
-                            acc
-                        }
-                    }
-                }
-            });
-        latest_regos.push(latest.clone())
-    }
-    latest_regos
+    Ok(reg_voting_power)
 }
 
 #[instrument]
 pub(crate) fn get_stake_address(
     stake_vkey_hex: &StakeVKey,
-    network_info: &NetworkInfo,
+    network: &NetworkInfo,
 ) -> Result<String> {
-    // Remove initial '0x' from string
-    if !stake_vkey_hex.starts_with("0x") {
-        warn!("stake_vkey_hex doesn't start with `0x`");
-    }
-
-    let stake_vkey_hex_only = stake_vkey_hex.trim_start_matches("0x");
+    let stake_vkey_hex = stake_vkey_hex.trim_start_matches("0x");
     // TODO support stake extended keys
-    if stake_vkey_hex_only.len() == 128 {
+    if stake_vkey_hex.len() == 128 {
```

Comment:
why is this `128`? And what does `stake_vkey` stand for?

Is the reference necessary for `hex::decode(&stake_vkey_hex)?;`

Reply:
Honestly not 100% certain, this was pretty much copied from the Haskell implementation

Reply:
I don't know that the reference is necessary, just a habit to pass references around